[quant][gpu][core] Added quantized linear operator in cudnn

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73959

Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.
Currently, int8 matrix multiplication is not supported in cudnn. The ETA for this support is in the first half of April 2022. As
a temporary workaround, we cast our int8 tensors to fp32 prior to matmul.

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D34824251](https://our.internmc.facebook.com/intern/diff/D34824251)

dzdang committed:
[quant][gpu][core] Added quantized linear operator in cudnn

Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.

[ghstack-poisoned]

jerryzh168 commented:
Please add a `Test Plan` for this PR as well

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn [PR currently incomplete]"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.

Test plan:

Differential Revision: [D34824251](https://our.internmc.facebook.com/intern/diff/D34824251)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn [PR currently incomplete]"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.
Currently, int8 matrix multiplication is not supported in cudnn. The ETA for this support is in the first half of April 2022. As
a temporary workaround, we cast our int8 tensors to fp32 prior to matmul.

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D34824251](https://our.internmc.facebook.com/intern/diff/D34824251)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34824251).

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn [PR currently incomplete]"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.
Currently, int8 matrix multiplication is not supported in cudnn. The ETA for this support is in the first half of April 2022. As
a temporary workaround, we cast our int8 tensors to fp32 prior to matmul.

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D34824251](https://our.internmc.facebook.com/intern/diff/D34824251)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34824251).

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn [PR currently incomplete]"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.
Currently, int8 matrix multiplication is not supported in cudnn. The ETA for this support is in the first half of April 2022. As
a temporary workaround, we cast our int8 tensors to fp32 prior to matmul.

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D34824251](https://our.internmc.facebook.com/intern/diff/D34824251)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.
Currently, int8 matrix multiplication is not supported in cudnn. The ETA for this support is in the first half of April 2022. As
a temporary workaround, we cast our int8 tensors to fp32 prior to matmul.

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D34824251](https://our.internmc.facebook.com/intern/diff/D34824251)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34824251).

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.
Currently, int8 matrix multiplication is not supported in cudnn. The ETA for this support is in the first half of April 2022. As
a temporary workaround, we cast our int8 tensors to fp32 prior to matmul.

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D34824251](https://our.internmc.facebook.com/intern/diff/D34824251)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34824251).

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.
Currently, int8 matrix multiplication is not supported in cudnn. The ETA for this support is in the first half of April 2022. As
a temporary workaround, we cast our int8 tensors to fp32 prior to matmul.

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D34824251](https://our.internmc.facebook.com/intern/diff/D34824251)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34824251).

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.
Currently, int8 matrix multiplication is not supported in cudnn. The ETA for this support is in the first half of April 2022. As
a temporary workaround, we cast our int8 tensors to fp32 prior to matmul.

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D34824251](https://our.internmc.facebook.com/intern/diff/D34824251)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34824251).

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Linear.cpp:
```diff
@@ -0,0 +1,345 @@
+// currently we only support int8 symmetric (zero_point = 0 for inputs and output) quantized linear op
+// We implement relu(act_int8 * transpose(w_int8) + [bias_fp32/(act_scale * w_scale] ) * ( act_scale * w_scale / out_scale )
+// which requires 5 cudnn ops (1 matmul, 2 multiplication, 1 add, and 1 relu ops)
+// matmul op: linear_op
+// Multiplication ops: rhs_mult_op, requant_op
+// Addition op: add_op
+// Relu op: relu_op
```
maybe we can write down the expressions for each of these ops as well?

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Linear.cpp:
```diff
@@ -0,0 +1,345 @@
+    return;
+  }
+  at::Tensor linear_output = at::empty(quantized_output.sizes(), at::device(at::kCUDA).dtype(at::kFloat));
+  // TODO: combine empty & fill_ using full_like or full
```
good point, maybe we can do it in this PR?

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Linear.cpp:
```diff
@@ -0,0 +1,345 @@
+    // we need to add trailing dimensions in order to properly broadcast bias, otherwise broadcast_to will fail.
+    // the number of trailling dimensions is quantized_output.dim() - 2, so the new size of the broadcast_bias
+    // becomes quantized_output.dim() - 2 + 1. nothing needs to be done for the leading dimensions
+    std::vector<int64_t> new_size(quantized_output.dim() - 1, 1);
```
1. is the call `ndim`?
2. why this is `quantized_output.dim() - 1`? shouldn't it match the dimension of quantized_output?

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Linear.cpp:
```diff
@@ -0,0 +1,345 @@
+  }
+  // cudnn expects tensors to be at least 3D. act is currently 2D. we will create a 3D view
+  // cudnn expects leading dimensions to be the dummy dimensions
+  std::vector<int64_t> new_sizes(3, 1);
+  new_sizes.back() = orig_weight.size(0);
+  new_sizes[1] = orig_weight.size(1);
```
this looks a bit unintuitive, maybe we can first transpose the weight, and use the dimension of transposed weight to create a 3d view sizes

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Linear.cpp:
```diff
@@ -0,0 +1,345 @@
+  auto weight_fp = weight_transposed.int_repr().to(at::kFloat);
+
+  auto run = [&](cudnn_frontend::ManagedOpaqueDescriptor plan_desc) {
+    auto workspace_size = 0;
```
I feel in general we have a lot of bolierplate code, maybe we can think about creating some helper functions or easier abstractions to make this simpler, this will be helpful when we have more ops in cudnn

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Linear.cpp:
```diff
@@ -0,0 +1,345 @@
+      // .setbMatDesc(cudnn_utils::getTensorDescriptor(orig_weight.sizes(), orig_weight.strides(), CUDNN_DATA_FLOAT, 'w', key.weight_alignment))
+      .setbMatDesc(cudnn_utils::getTensorDescriptor(weight_fp.sizes(), weight_fp.strides(), CUDNN_DATA_FLOAT, 'w', key.weight_alignment))
+      .setcMatDesc(cudnn_utils::getTensorDescriptor(linear_output, 'y', key.output_alignment))
+      .setmatmulDesc(getLinearDescriptor(CUDNN_DATA_FLOAT)) // is this right? should it be float?
```
I remember we have a table for the descriptor data type, maybe we can implement that in a function: getting descriptor data type from input data type

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Linear.cpp:
```diff
@@ -0,0 +1,345 @@
+      .build());
+  }
+
+  // relu_op computes relu(act_int8 * w_int8 + [bias_fp32/(act_scale * w_scale)]) / (out_scale / (act_scale * w_scale))
```
`requant_op`?

jerryzh168 reviewed (approved):
Looks good, had some nit comments inline

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.
Currently, int8 matrix multiplication is not supported in cudnn. The ETA for this support is in the first half of April 2022. As
a temporary workaround, we cast our int8 tensors to fp32 prior to matmul.

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D34824251](https://our.internmc.facebook.com/intern/diff/D34824251)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34824251).

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.
Currently, int8 matrix multiplication is not supported in cudnn. The ETA for this support is in the first half of April 2022. As
a temporary workaround, we cast our int8 tensors to fp32 prior to matmul.

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D34824251](https://our.internmc.facebook.com/intern/diff/D34824251)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34824251).

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.
Currently, int8 matrix multiplication is not supported in cudnn. The ETA for this support is in the first half of April 2022. As
a temporary workaround, we cast our int8 tensors to fp32 prior to matmul.

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D34824251](https://our.internmc.facebook.com/intern/diff/D34824251)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Added quantized linear operator in cudnn"


Summary:
This PR is similar to https://github.com/pytorch/pytorch/pull/70622, but for the linear operator.
Unlke PR 70622, this implementations directly uses packed parameters, rather than a refactorization, as was done for the conv operator,
and also directly implements bias & relu.
Currently, int8 matrix multiplication is not supported in cudnn. The ETA for this support is in the first half of April 2022. As
a temporary workaround, we cast our int8 tensors to fp32 prior to matmul.

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D34824251](https://our.internmc.facebook.com/intern/diff/D34824251)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34824251).
